text/0000-unwinding.md

```diff
@@ -0,0 +1,116 @@
+is impossible to prevent.
+
+# Detailed design
+
+## Semantics
+
+The current Rust semantics are that failure leaves the task in an invalid state and therefore
+continued execution within that task is undefined behaviour. "Task" currently refers to threads,
+however this RFC proposes to extend the definition of a task to include the entry into Rust code
+from non-Rust code. As such, the task ends when control returns to original non-Rust code.
+
+This extension essentially produces "nested tasks", where the thread is one task, then each transfer
+of control from non-Rust code to Rust code is another. Unwinding out of a task is undefined
+behaviour, meaning that the current behaviour regarding unwinding into non-Rust code is preserved.
+However, the extended definition of "task" allows for unwinding to be caught at the entry point of
+the Rust code. From there a user may decide to return an error code or abort the process.
+
+Attempts to catch unwinding at any point other than a task entry point is considered undefined behaviour.
+
+## Implementation
+
+Rust already has a significant amount of code for handling unwinding. However, it is currently not
+safe to use nested invocations of the `try` function. The first part of the implementation would be
+to improve the unwinding system to allow for nested calls to the try function.
+
+The `try` function itself doesn't need changing. The current signature is `fn try<F:FnOnce()>(f: F)
+-> Result<(), Box<Any + Send>>`, which provides means for the caller to know if the invocation was
+successful. The function would remain unsafe.
+
+An example:
+
+```rust
+extern "C" fn foo() -> c_int {
+    let mut ret = 0;
+    let res = try(|| {
+	    ret = inner_call();
+    });
+
+    match res {
+	    Err(_) => return -1,
+		Ok(_) => return ret,
+	}
+}
+```
+
+As a fallback, if the user does not catch the unwinding themselves, the process is aborted. This is
+the only thing can reasonably be done if we attempt to unwind into code that likely does not know
+how to handle it.
+
+## Edge Cases and Other Miscellany
+
+Sometime you need to be able to unwind from a non-Rust ABI function, a particularly relevant
+instance of this is the `rust_try` function itself, as it uses the C ABI. With the above semantics,
+this would cause the process to be aborted, not what would be wanted. Therefore a new attribute
+`#[can_unwind]` is used to indicate that the function in question may unwind, and that this
+unwinding is expected.
+
+As a complement to `#[can_unwind]`, another attribute `#[unsafe_no_unwind]` (named such as it is
+unsafe when used incorrectly) can be used to mark functions as specifically not unwinding. This is
+primarily for performance reasons in cases where the function cannot be inferred to not unwind. The
+primary use case for this attribute is the `oom` function that aborts the process in the case of a
+out-of-memory condition.
```

Comment:
Out of curiosity, I think I remember hearing that code size benefited from this change, do you have some examples or numbers that this could cite to show the benefits that it reaps?